### PR TITLE
[SHELL32] Refactor DefView column handling

### PR DESCRIPTION
**Pull request by whindsaks**
This refactor decouples the previous assumption that the ListView column index is the same thing as the IShellFolder column index. By doing this, support for turning columns on or off becomes possible. The `SHCOLSTATE_ONBYDEFAULT` and `SHCOLSTATE_HIDDEN` flags are also respected.

I also fixed the "Arrange icons by" submenu (JIRA issue: [CORE-18526](https://jira.reactos.org/browse/CORE-18526)).
Because the existing code unfortunatly used the entire menu id range allocated for IShellView (`FCIDM_SHVIEWFIRST`..`FCIDM_SHVIEWLAST`) when dealing with IContextMenu, I did have to modify that code slightly because I needed some menu item ids.

I moved two of the existing menu helper functions because I needed to call one of them higher up in the file.

The "More" menu item has been disabled in this PR because the "Choose Details" dialog has not been implemented yet. The code was however coded to support it, all that is missing is the dialog. This dialog is the only way for keyboard-only users to modify the columns and it's the only way to enable columns marked with `SHCOLSTATE_SECONDARYUI`. I can implement the dialog and its logic in another PR if it is deemed useful.

DV col. and arrange menus

### CODE REVIEW

File: dll/win32/shell32/CDefView.cpp [hunk 1]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+        UINT count = DPA_GetPtrCount(m_ListToFolderColMap);
+        if (ListCol < count)
+        {
+            HRESULT col = (INT)(INT_PTR) DPA_FastGetPtr(m_ListToFolderColMap, ListCol);
+            assert(col >= 0 && col == SHGetLVColumnSubItem(m_ListView.m_hWnd, ListCol));
+            return col;
+        }
+        else if (count)
+        {
+            TRACE("m_ListToFolderColMap cache miss while mapping %d!\n", ListCol);
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefView.cpp:892]:
```suggestion
            TRACE("m_ListToFolderColMap cache miss while mapping %d\n", ListCol);
```

File: dll/win32/shell32/CDefView.cpp [hunk 2]
```diff
@@ -1475,6 +1756,105 @@ HRESULT CDefView::CheckViewMode(HMENU hmenuView)
...
+        ::GetWindowRect(hWndHdr, &r);
+        pt.x = r.left + ((r.right - r.left) / 2);
+        pt.y = r.top + ((r.bottom - r.top) / 2);
+    }
+
+    HMENU hMenu = CreatePopupMenu();
+    if (!hMenu)
+        return 0;
+
+    for (UINT foldCol = 0;; ++foldCol)
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefView.cpp:1781]: Place the declaration at the beginning of the function, not within the for-loop.
Here, and in the other locations where you did it like that.

File: dll/win32/shell32/CDefView.cpp [hunk 3]
```diff
@@ -1884,6 +2271,13 @@ LRESULT CDefView::OnCommand(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHand
 
     TRACE("(%p)->(0x%08x 0x%08x %p) stub\n", this, dwCmdID, dwCmd, hwndCmd);
 
+    if (dwCmdID >= DVIDM_ARRANGESORT_FIRST && dwCmdID <= DVIDM_ARRANGESORT_LAST)
+    {
+        UINT listCol = (UINT) GetMenuItemDataById(m_hMenuArrangeModes, dwCmdID);
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefView.cpp:2276]:
```suggestion
        UINT listCol = (UINT)GetMenuItemDataById(m_hMenuArrangeModes, dwCmdID);
```

**Review by JoachimHenze — CHANGES_REQUESTED** [review]

File: dll/win32/shell32/CDefView.cpp (hunk 2 above)
Comment (JoachimHenze) [reply dll/win32/shell32/CDefView.cpp:1781]: Nvm, I overlooked that we are cpp code here. It is ok how you did that.

File: dll/win32/shell32/CDefView.cpp (hunk 3 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:2276]: Where in the coding style document does it say this is required?

Comment (JoachimHenze) [reply dll/win32/shell32/CDefView.cpp:2276]: It is not enforced in the coding-styles, but it usually saves line-width. It's a suggestion, not more.

File: dll/win32/shell32/CDefView.cpp [hunk 4]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+{
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_SUBITEM;
+    if (!ListView_GetColumn(List, Col, &lvc))
+        return E_FAIL;
+    else
+        return lvc.iSubItem;
+}
+
+HRESULT CDefView::MapFolderToList(UINT FoldCol)
```
Comment (yagoulas) [inline dll/win32/shell32/CDefView.cpp:864]: Rename this to MapFolderColumnToListColumn

File: dll/win32/shell32/CDefView.cpp [hunk 5]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+    {
+        HRESULT r = SHGetLVColumnSubItem(m_ListView.m_hWnd, i);
+        if ((UINT) r == FoldCol)
+            return i;
+        else if (FAILED(r))
+            return r;
+    }
+}
+
+HRESULT CDefView::MapListToFolder(UINT ListCol)
```
Comment (yagoulas) [inline dll/win32/shell32/CDefView.cpp:877]: Rename this to MapListColumnToFolderColumn

File: dll/win32/shell32/CDefView.cpp [hunk 6]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+        }
+        else if (count)
+        {
+            TRACE("m_ListToFolderColMap cache miss while mapping %d\n", ListCol);
+        }
+    }
+    return SHGetLVColumnSubItem(m_ListView.m_hWnd, ListCol);
+}
 
+HRESULT CDefView::GetDetailsOfFromFolder(PCUITEMID_CHILD pidl, UINT FoldCol, SHELLDETAILS &sd)
```
Comment (yagoulas) [inline dll/win32/shell32/CDefView.cpp:898]: Rename this to GetDetailsByFolderColumn

File: dll/win32/shell32/CDefView.cpp [hunk 7]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+#if 0 // TODO
+    if (FAILED(hr))
+    {
+        FIXME("Try SFVM_GETDETAILSOF\n");
+    }
+#endif
+    return hr;
+}
+
+HRESULT CDefView::GetDetailsOfFromList(PCUITEMID_CHILD pidl, UINT ListCol, SHELLDETAILS &sd)
```
Comment (yagoulas) [inline dll/win32/shell32/CDefView.cpp:920]: Rename this to GetDetailsByListColumn

File: dll/win32/shell32/CDefView.cpp [hunk 8]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+            ListCol = 0x7fffffff;
+    }
+    if (Insert)
+        ListView_InsertColumn(m_ListView.m_hWnd, ListCol, &lvc);
+    else
+        ListView_SetColumn(m_ListView.m_hWnd, ListCol, &lvc);
+    return S_OK;
+}
+
+HRESULT CDefView::LoadColumns(UINT *pColList, UINT ColListCount)
```
Comment (yagoulas) [inline dll/win32/shell32/CDefView.cpp:962]: What is the purpose of these parameters? Do you plan for them to be used in the future and in what way?

File: dll/win32/shell32/CDefView.cpp [hunk 9]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+        if (FAILED(hr))
+        {
+            if (!pColList)
+                hr = S_OK; // No more items, we are done
+            break;
+        }
+        ++newColCount;
     }
-    else
+    for (i = newColCount; i < oldColCount; ++i)
```
Comment (yagoulas) [inline dll/win32/shell32/CDefView.cpp:1010]: Will this ever be used? So far this function is called only on initialization and as far as I understand there are no columns in the list at this point.

File: dll/win32/shell32/CDefView.cpp (hunk 8 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:962]: It is how the More dialog will specify the column list.

File: dll/win32/shell32/CDefView.cpp (hunk 9 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1010]: When the More dialog specifies fewer columns than what the list currently has.

**Comment by whindsaks** [comment]
Go away bot

File: dll/win32/shell32/CDefView.cpp [hunk 10]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+
     m_hMenuArrangeModes = CreateMenu();
+    LoadColumns();
+    return TRUE;
+}
+
+/**********************************************************
+* Column handling
+*/
+static HRESULT SHGetLVColumnSubItem(HWND List, UINT Col)
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:854]: Since I would expect that this function would mostly succeed, and return actual indices instead of an error, I would suggest you make it have `INT` return type (or whatever the `LVCOLUMN::iSubItem` type is), instead of "HRESULT" (which means "an error code").
And then, just cast the `E_FAIL` to `INT` or whatever.

Or, make the function return an `UINT` and in case of error, make it return `UINT_MAX`, like what has been done in e.g. the ReallyGetMenuItemID function.

File: dll/win32/shell32/CDefView.cpp [hunk 11]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+        return lvc.iSubItem;
+}
+
+HRESULT CDefView::MapFolderColumnToListColumn(UINT FoldCol)
+{
+    // This function is only called during column management, performance is not critical.
+    for (UINT i = 0;; ++i)
+    {
+        HRESULT r = SHGetLVColumnSubItem(m_ListView.m_hWnd, i);
+        if ((UINT) r == FoldCol)
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:870]: So that here, the UINT cast wouldn't be needed.

File: dll/win32/shell32/CDefView.cpp [hunk 12]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+            return i;
+        else if (FAILED(r))
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:872]:
```suggestion
        else if (r == E_FAIL)
```

File: dll/win32/shell32/CDefView.cpp [hunk 13]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+{
+    // This function is called every time a LVITEM::iSubItem is mapped to
+    // a folder column (calls to GetDetailsOf etc.) and should be fast.
+    if (m_ListToFolderColMap)
+    {
+        UINT count = DPA_GetPtrCount(m_ListToFolderColMap);
+        if (ListCol < count)
+        {
+            HRESULT col = (INT)(INT_PTR) DPA_FastGetPtr(m_ListToFolderColMap, ListCol);
+            assert(col >= 0 && col == SHGetLVColumnSubItem(m_ListView.m_hWnd, ListCol));
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:887]: Better keep `col` to an `INT` or whatever. And have `CDefView::MapListColumnToFolderColumn` also return `INT`.

File: dll/win32/shell32/CDefView.cpp [hunk 14]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+    return hr;
+}
+
+HRESULT CDefView::LoadColumn(UINT FoldCol, UINT ListCol, BOOL Insert)
+{
+    WCHAR buf[MAX_PATH];
+    SHELLDETAILS sd;
+    HRESULT hr;
+
+    sd.str.uType = !STRRET_WSTR; // Protect ourself from broken implementations
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:935]: "not" STRRET_WSTR ? What does that mean?

File: dll/win32/shell32/CDefView.cpp [hunk 15]
```diff
@@ -1097,16 +1376,22 @@ HRESULT CDefView::FillList()
...
     {
-        FIXME("no m_pSF2Parent\n");
+        m_sortInfo.bIsAscending = TRUE;
+        sortCol = 0; // In case the folder does not know/care
+        if (m_pSF2Parent)
+        {
+            ULONG foldSortCol = sortCol;
+            HRESULT hr = m_pSF2Parent->GetDefaultColumn(NULL, &foldSortCol, NULL);
+            if (SUCCEEDED(hr))
+                hr = MapFolderColumnToListColumn(foldSortCol);
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:1389]: It would be better here (in regards to my previous comments) to use the `foldSortCol` variable here. (Maybe using a different name because it's not "old" now...)

File: dll/win32/shell32/CDefView.cpp [hunk 16]
```diff
@@ -1475,6 +1756,105 @@ HRESULT CDefView::CheckViewMode(HMENU hmenuView)
     return S_OK;
 }
 
+LRESULT CDefView::DoColumnContextMenu(LPARAM lParam)
+{
+    const UINT maxItems = 15; // Feels about right
+    const UINT idMore = 0x1337;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:1762]: Why these hardcoded values?

**Review by HBelusca — COMMENTED** [review]
Some stuff to check; the PR needs to be rebased on top of master too.

File: dll/win32/shell32/CDefView.cpp (hunk 10 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:854]: Hresult is an int.

File: dll/win32/shell32/CDefView.cpp (hunk 12 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:872]: It might return another error in the future, who knows. That is why those macros exit.

File: dll/win32/shell32/CDefView.cpp (hunk 14 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:935]: It means exactly this, initialize to something that is not wstr.

This code calls into unknown (possibly 3rd party) code and we don't trust it. The wstr type needs to be free'ed by us. This handles the case where the callee is broken but still returns S_OK and our stack garbage happened to be wstr.

File: dll/win32/shell32/CDefView.cpp (hunk 16 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1762]: I don't remember the details anymore but I believe Windows also limits the number of items it will display in this menu. Just like the comment says, 15 feels like the maximum that looks decent. If you want a less used column, you would have to click the "more" item (that is in another PR and is not going to do much until IShellFolder2 is fully implemented).

File: dll/win32/shell32/CDefView.cpp [hunk 17]
```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
...
+        return hr;
+    hr = StrRetToStrNW(buf, _countof(buf), &sd.str, NULL);
+    if (FAILED(hr))
+        return hr;
+
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_TEXT | LVCF_FMT | LVCF_WIDTH | LVCF_SUBITEM;
+    lvc.pszText = buf;
+    lvc.fmt = sd.fmt;
+    lvc.cx = sd.cxChar * 8; // FIXME: Font size? DPI?
```
Comment (DarkFire01) [inline dll/win32/shell32/CDefView.cpp:947]: Might be outta scope but it would be nice to know what this is for sure?

File: dll/win32/shell32/CDefView.cpp (hunk 16 above)
Comment (DarkFire01) [reply dll/win32/shell32/CDefView.cpp:1762]: What's idMore then I mean it's 0x1337 so maybe you chose that on purpose lol

File: dll/win32/shell32/CDefView.cpp (hunk 17 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:947]: If I copy zipfldr.dll from XP to ROS, it reports 

```
Name:20
Type:15
Packed Size:10
Has a password:8
Size:8
Ratio:6
Date:25
```
for its `SHELLDETAILS::cxChar` values to our DefView. Mapping these to pixels with `TEXTMETRIC::tmAveCharWidth` does not match XP for whatever reason. `GetTextExtentPointW` gets very close however.

XP and ROS

File: dll/win32/shell32/CDefViewBckgrndMenu.cpp [hunk 18]
```diff
@@ -256,10 +256,6 @@ CDefViewBckgrndMenu::InvokeCommand(LPCMINVOKECOMMANDINFO lpcmi)
     case FCIDM_SHVIEW_SMALLICON:
     case FCIDM_SHVIEW_LISTVIEW:
     case FCIDM_SHVIEW_REPORTVIEW:
-    case 0x30: /* FIX IDS in resource files */
-    case 0x31:
-    case 0x32:
-    case 0x33:
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefViewBckgrndMenu.cpp:262]: Is this change related/dependant on your other changes in this PR? Or is this unrelated cleanup which could be applied also without all your other changes?

File: dll/win32/shell32/CDefView.cpp [hunk 19]
```diff
@@ -783,40 +865,243 @@ void CDefView::UpdateListColors()
...
+    if (!chavewidth)
+        chavewidth = 6; // 6 is a reasonable default fallback
+
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_TEXT | LVCF_FMT | LVCF_WIDTH | LVCF_SUBITEM;
+    lvc.pszText = buf;
+    lvc.fmt = sd.fmt;
+    lvc.cx = sd.cxChar * chavewidth; // FIXME: DPI?
+    lvc.iSubItem = FoldCol; // Used by MapFolderColumnToListColumn & MapListColumnToFolderColumn
+    if ((int) ListCol == -1)
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefView.cpp:985]:
```suggestion
    if ((int)ListCol == -1)
```
It helps seeing that the cast is applied first, before the == is evaluated. Nit: Actually all spaces after casts are questionable, but this one here hurts the readers interpretation most. Otherwise your PR lgtm.

**Review by JoachimHenze — APPROVED** [review]

File: dll/win32/shell32/CDefViewBckgrndMenu.cpp (hunk 18 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefViewBckgrndMenu.cpp:262]: They are related to sorting but are totally broken.

The new sort menu ids are DVIDM_ARRANGESORT_FIRST to DVIDM_ARRANGESORT_LAST and now outside the range passed to QueryContextMenu.

File: dll/win32/shell32/CDefView.cpp (hunk 16 above)
Comment (DarkFire01) [reply dll/win32/shell32/CDefView.cpp:1775]: Good enough for me.

File: dll/win32/shell32/CDefView.cpp (hunk 14 above)
Comment (HBelusca) [reply dll/win32/shell32/CDefView.cpp:967]: Please explain that in a comment in the code. Because this is not something commonly seen in code bases.

File: dll/win32/shell32/CDefView.cpp (hunk 15 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1407]: It is not "old" (f-old), I'll change the name to make it clear.
